enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,756 @@
+In the future, where we have both Machine API and Cluster API authoritative MachineSets, the Autoscaler must be updated to ensure it scales the authoritative MachineSet.
+
+Since every Machine API MachineSet will be reflected into Cluster API, the autoscaler can be updated to reconcile only the Cluster API MachineSets.
+When a scaling decision is made, it must then update the authoritative MachineSet to the correct scale.
+
+To do this, there are two possible approaches to consider:
+- Look up the Machine API MachineSet mirror to determine the authoritative API and update accordingly
+  - This would be a carry patch to the autoscaler, which, may cause maintenance toil over time
+- Allow the Autoscaler to update the Cluster API MachineSet, but allow the synchronisation controller to identify the update has come from the Autoscaler, and mirror it to the authoritative MachineSet as appropriate
+  - This could be achieved by observing the managed fields property of the replicas field and determine which actor updated it.
+  If the Autoscaler last changed the replica count, it should be mirrored back to the Machine API MachineSet if it is authoritative.
+  - This creates an exception to the rules outline above and adds complication to the conversion, but, may be less of a burden than carrying a patch in the autoscaler
+
+Both avenues above will need to be explored further before a decision is made on how the Autoscaler will handle mixed MachineSet clusters.
+
+#### Validation of the authoritative API
+
+The authoritative API has two valid values, `machine.openshift.io` and `cluster.x-k8s.io`. Any other value is an error.
+To prevent invalid values, a new validating admission webhook will be introduced to validate the value of the annotation.
+
+This webhook will be configured to accept resources from the Machine API group and check the annotation value is permitted.
+Any invalid value will be rejected by the webhook.
+
+In addition to this, the webhook will validate, on a change of authoritative API, that the synchronisation is up to date.
+If the value of the `machine.openshift.io/synchronised-generation` annotation does not match the current generation,
+the change will be rejected and requested to be retried once the synchronisation has occurred.
+
+### Risks and Mitigations
+
+#### Synchronisation logic will need to be solid before it gets into the hand of customers
+
+To ensure that we do not break customer clusters, we need to make sure that our logic that converts the resources between Machine API and Cluster API is thoroughly tested.
+
+To ensure we don’t break customers we will write extensive tests for the conversion, primarily, we expect this to be done via unit testing, however, we will still need to test these conversions in E2E however, since the semantics of a configuration option may differ between the two APIs.
```

Comment:
Similar to having a library for filtering reconciles, could the logic for conversion reside in a library that provides a wrapper CLI?

This way, we can have a suite of YAML fixtures that we can do conversions on in CI without having to spin up a full cluster every time.

I'm envisioning something like

```
func MachineSetToMAPI(*capi.MachineSet) *mapi.MachineSet
func MachineSetToCAPI(*mapi.MachineSet) *capi.MachineSet
...
func <Type>To<APIGroup>...
```

Of course, we should still have extensive e2e tests, but this may be a way to make the inner loop of testing conversion logic faster.

Reply:
I had assumed we could add a unit testing suite for all of the conversions which would allow us to internally test these kinds of things. A CLI was floated in the past as an option to allow users to convert resources they have away from clusters. For example gitops style resources.

If we still think there's value in that I don't think it would be difficult to implement a CLI and package it in a way similar to CCO do for their ccoctl tool